src/pie_utils/processor/document/regex_partitioner.py

```diff
@@ -0,0 +1,165 @@
+    partition label. The span of the partition ranges between two matched pattern. It should be
+    noted that none of the partitions overlap.
+
+    Besides the regular partitioning of the document, there are some other cases listed below:
+    1. Document can be converted into single partition with initial_partition_label if skip_initial_partition is False
+       and label_whitelist is empty list or label_group_id is None.
+    2. If only initial part of the document is required as a partition then set label_group_id and label_whitelist as
+       None and skip_initial_partition as False.
+    3. There will be no partitions if
+        a. matcher could not find pattern in document text
+        b. label_whitelist is empty list and skip_initial_partition is True
+        c. label_group_id and label_whitelist is None and skip_initial_partition is True
+        d. label_group_id is None and skip_initial_partition is False
+
+    :param document: A Document that is to be partitioned
+    :param matcher: A method that is used to match a pattern in the document text and return an iterator yielding the
+                    Match objects.
+    :param label_whitelist: A list of labels which are allowed to form a partition. (the default value is None)
+    :param label_group_id: An integer value to select suitable match group from Match object. (the default value is None)
+    :param skip_initial_partition: A boolean value that prevents initial partition to be saved in the document. (the
+                                default value is False)
+    :param initial_partition_label: A string value used as a partition label for initial partition. This is only used when
+                                skip_initial_partition is False.
+    """
+    previous_start = previous_label = None
+    if not skip_initial_partition:
+        previous_start = 0
+        previous_label = initial_partition_label
+    for match in matcher(document.text):
+        if label_group_id is not None:
+            start = match.start(label_group_id)
+            end = match.end(label_group_id)
+            label = document.text[start:end]
+        else:
+            label = None
+        if label_whitelist is None or label in label_whitelist:
+            if previous_start is not None and previous_label is not None:
+                end = match.start()
+                span = LabeledSpan(start=previous_start, end=end, label=previous_label)
+                yield span
+
+            previous_start = match.start()
+            previous_label = label
+
+    if previous_start is not None and previous_label is not None:
+        end = len(document.text)
+        span = LabeledSpan(start=previous_start, end=end, label=previous_label)
+        yield span
+
+
+class RegexPartitioner(WithStatistics):
+    """RegexPartitioner partitions a document into multiple parts using a regular expression.
```

Comment:
important information to add: the matched string is included in the parts (at the beginning)!

Reply:
also, at least refer to the method `get_partitions_with_matcher` for further information.